### PR TITLE
Fix LockContext cancellation deadlock; harden queue; upgrade to Go 1.26

### PR DESCRIPTION
**Pull request by neilotoole**
## TL;DR

- **Critical fix**: resolves a reproducible deadlock in `Mutex.LockContext` when `ctx` cancellation races with `Unlock`. Root cause: unbuffered pooled handoff channel. Fixed by `buffered(1) + select-with-default-panic` self-enforcing invariant.
- **Supporting hardening**: `list.remove` double-pool guard, `Unlock` recover-safety (via `locked bool` refactor + deferred inner unlock), waiter methods encapsulating the protocol, explicit `noCopy`, generics dropped from the internal `list`, `cur int64` → `locked bool`.
- **Twenty-one new tests** covering `LockContext`, FIFO invariants, queue corner cases, pool invariants, chaos, goroutine-leak floor, recover-safety, and concurrent double-unlock.
- **Three new benchmarks** covering the `LockContext` surface, plus a full re-run of the existing suite on Go 1.26.
- **Extensive docs**: README gains "When to use", "How it works", "Testing and correctness", "Requirements" sections; package doc picks up godoc headings; method/field comments tightened throughout.
- **Toolchain & CI**: `go.mod` to 1.26; test step runs with `-race`; GitHub Actions bumped (`checkout@v6`, `setup-go@v6`, `golangci-lint-action@v9`); golangci-lint config migrated v1→v2 (pinned `v2.12.2`); `push` trigger scoped to `master` to stop duplicate PR runs; added concurrency cancellation and least-privilege permissions.
- **Two rounds of AI-assisted code review** were run on the PR before this description; their findings are incorporated into commits 1–16.

Origin: [r/golang thread 1bek63o](https://www.reddit.com/r/golang/comments/1bek63o/fifomu_mutex_with_fifo_lock_acquisition/), specifically u/skeeto's [comment](https://www.reddit.com/r/golang/comments/1bek63o/comment/kuv38y0/) on FIFO fairness.

---

### Reproducer (pre-fix)

```go
for i := 0; i < 5000; i++ {
    var mu fifomu.Mutex
    mu.Lock()
    ctx, cancel := context.WithCancel(context.Background())
    cancel() // already canceled
    done := make(chan error, 1)
    go func() { done <- mu.LockContext(ctx) }()
    runtime.Gosched()
    mu.Unlock()

    select {
    case <-done:
    case <-time.After(3 * time.Second):
        panic(\"deadlock\")
    }
}
```

### Trace (pre-fix)

```
goroutine 6 [chan send]:
  fifomu.(*Mutex).notifyWaiters    fifomu.go:159   <- w <- struct{}{}
  fifomu.(*Mutex).Unlock           fifomu.go:138   <- still holding m.mu

goroutine 34 [sync.Mutex.Lock]:
  sync.(*Mutex).Lock
  fifomu.(*Mutex).LockContext      fifomu.go:89    <- m.mu.Lock() in ctx.Done branch
```

### Why

`notifyWaiters` sent on an **unbuffered** channel while holding `m.mu`. If `LockContext`'s select picked `ctx.Done()` instead of `<-w`, the send never synchronized, so the Unlock goroutine parked on `sendq` while still holding `m.mu`. The canceling `LockContext` then blocked on `m.mu.Lock()` to reach its inner cleanup select. Deadlock — and every subsequent caller piled up behind `m.mu`.

The upstream (`golang.org/x/sync/semaphore.Weighted`) doesn't have this bug because it calls `close(w.ready)` — non-blocking for all receivers. `fifomu` replaced `close` with `w <- struct{}{}` to allow pool reuse, which is precisely what introduced the deadlock.

### Fix

Pooled handoff channels are now **buffered size 1**, and the send is wrapped in `select { case w <- struct{}{}: default: panic(...) }`. The pool invariant (waiters return empty) means the send never blocks; if anyone ever violates the invariant, we panic loudly instead of reintroducing the deadlock. The panic is caught by `TestNotifyWaiters_PanicsOnViolatedInvariant`.

---

### Queue hardening
- **`list.remove` early-return** when the element is not in this list. Previously a double-remove would `Put` the same element into `elementPool` twice, letting two concurrent `Get` calls receive the same pointer. `TestList_RemoveIsIdempotent`, `TestList_RemoveDoesNotDoublePool`, and `TestList_RemoveClearsFields` lock this in.
- **Cancel-path pool return**: `LockContext`'s cancel handler now returns the waiter to the pool on both the `<-w` and default branches (previously leaked on `default`).
- **`notifyWaiters` simplified** from a loop to a single branch (binary mutex never signals more than one waiter per call). The dead `m.locked` guard was dropped after the `locked bool` refactor.
- **`Unlock` uses `defer m.mu.Unlock()`** so any panic inside (including the invariant-violation panic from `signal`) releases the inner mutex instead of wedging the whole package.

### Type cleanup
- **`cur int64` → `locked bool`**. Expresses the binary invariant in the type. As a side effect, `Unlock` now check-before-mutates, which means a recovered double-unlock panic leaves the Mutex **usable** (unlike `sync.Mutex`, which corrupts state and is recover-hostile). Verified by `TestMutex_RecoverFromDoubleUnlockPanic`.
- **Explicit `noCopy` field** on `Mutex`, independent of the stdlib's implementation details.
- **Generics dropped** from the internal `list`. It only ever instantiated as `list[waiter]`, and the pool was already concrete. Collapsing shrinks the binary and removes dead abstraction.
- **`element.Value` → `element.value`** (unexported) and **`list.len uint` → `int`** (conventional; no longer load-bearing against underflow after the `remove` early-return).
- **`waiter` gains methods**: `signal()` / `wait()` / `tryReceive()`, encapsulating the protocol invariants (including the self-enforcing send's panic-on-violation) in one place. `list.pushBack` (a trivial wrapper) was deleted after the concretization.

### Tests (zero → twenty-one for the new surface)

External tests (`fifomu_context_test.go`, `fifomu_invariants_test.go`, `fifomu_chaos_test.go`, `main_test.go`):

| Area | Tests |
|---|---|
| `LockContext` fast path | `_AlreadyCanceledFastPath` |
| `LockContext` slow path | `_AlreadyCanceledSlowPath`, `_BlockedCancel`, `_DeadlineExceeded`, `_ContextCause` |
| Race regression | `_CancelRaceRegression` — deadlocks within seconds if the fix is reverted |
| Race outcomes | `_AcquireAndCancelOutcomes` — exercises the cancel/unlock race and logs the acquired/canceled split; asserts every outcome is well-formed and no iteration deadlocks. Which side wins is hardware-dependent, so neither count is required to be nonzero (the deterministic canceled path is covered by `_BlockedCancel`) |
| Stress | `_HammerConcurrent`, `TestMutex_Chaos` (20 goroutines × 2s with mutual-exclusion counter) |
| FIFO | `TestMutex_FIFOOrdering`, `TestLockContext_FIFOAmongContextWaiters`, `TestMutex_MixedQueueLockAndLockContext` |
| Cancel corner | `TestMutex_MiddleOfQueueCancel` |
| API semantics | `TestMutex_UnlockFromDifferentGoroutine`, `TestMutex_UnlockPanicMessage`, `TestMutex_RecoverFromDoubleUnlockPanic`, `TestMutex_ConcurrentDoubleUnlock`, `TestTryLock_RefusesWhileWaiterQueued` |
| Goroutine leak floor | `TestMutex_NoGoroutineLeakAfterCancel` plus `goleak.VerifyTestMain` |

Internal tests (`internal_test.go`):

| Area | Tests |
|---|---|
| List invariants | `TestList_RemoveIsIdempotent`, `TestList_RemoveDoesNotDoublePool`, `TestList_RemoveClearsFields` |
| Invariant enforcement | `TestNotifyWaiters_PanicsOnViolatedInvariant` |

Deterministic FIFO tests use a test-only `fifomu.WaitersLen(*Mutex)` accessor (via `export_test.go`) instead of `time.Sleep` gates, eliminating flakiness on loaded CI.

### Benchmarks

New `LockContext` coverage (M1 Max, Go 1.26, `GOMAXPROCS=10`):

| Benchmark | ns/op | allocs/op | Comment |
|---|---:|---:|---|
| `LockContext_FastPath` | 3.457 | 0 | parity with `Lock` within noise |
| `LockContext_Contended` | 191.8 | 0 | ~18% over contended `Lock` (extra select arm) |
| `LockContext_Cancel` | 88.34 | 0 | slow-path cancel, alloc-free |

Existing `BenchmarkMutex*` suite re-run on Go 1.26. Post-refactor performance is within noise of pre-refactor, confirming B1–B4 are perf-neutral.

### Docs

README gains four new sections: **When to use** (motivating examples, streamcache pattern, rate-limited queues), **How it works** (buffered(1) handoff, pooled waiters, no spinning), **Testing and correctness** (race, goleak, chaos, regression), **Requirements** (Go 1.26, deps). A benchmark-name glossary explains `Slack`/`NoSpin`/`Spin`/etc.

Package doc uses godoc-style section headings (`# When to use`, `# FIFO guarantee`, `# How it works`, `# When to prefer sync.Mutex`) and now documents the `TryLock` divergence at the package level.

Method and field comments tightened throughout: `Unlock` documents recover-safety specifically (scoped to the double-unlock panic); `signal` explains the deadlock history in two paragraphs (punchline not buried); `list.go` documents `pushBackElem`, the `element.list` back-pointer, and `lazyInit`.

### Toolchain
- `go.mod`: `go 1.26`. The library itself uses only pre-1.20 features; `for b.Loop()` (1.24+) and `sync.WaitGroup.Go` (1.25+) are used in tests/benchmarks. The declared version is the tested floor.
- CI actions: `actions/checkout@v6`, `actions/setup-go@v6`, `golangci/golangci-lint-action@v9` (pinned to golangci-lint `v2.12.2`), resolving the Node 20 deprecation on the older versions. Test step runs with `-race`; benchmark step pinned to `go test -bench=. -benchmem -run=^$`.
- Lint config: `.golangci.yml` migrated from golangci-lint v1 to v2 — v1 is EOL with no go1.26 build, so it could no longer lint a `go 1.26` module (`the Go language version (go1.24) used to build golangci-lint is lower than the targeted Go version (1.26)`). Migrated via the official `golangci-lint migrate`; `gomodguard` → `gomodguard_v2`; fixed a stale `gofumpt.module-path`; test-only idioms (unchecked pool-get assertions, `const N`, `runtime.GC()` in the reclamation test, the empty `TryLock` branch) excluded under `_test.go`.
- CI triggers: `push` scoped to `master` so PRs no longer run twice (a `push` set **and** a `pull_request` set); added `concurrency` cancellation of superseded runs and least-privilege `permissions: contents: read`.
- One new test-only dep: `go.uber.org/goleak`.

---

## Review history in this branch

Two rounds of AI-assisted code review were run on this PR via the parallel-agent review pattern (pr-review-toolkit: code-reviewer, comment-analyzer, pr-test-analyzer, silent-failure-hunter, type-design-analyzer). Findings were triaged and addressed as follow-up commits. Everything flagged as Critical or Important has been resolved; a handful of lower-criticality items were deferred and are noted below.

A subsequent GitHub Copilot review round flagged four items on the updated branch: three were fixed in `fb7e9a0` (the two `time.Sleep` enqueue waits → the deterministic `waitForWaiters` gate, and the README `x/sync` dependency label), and one was declined as a false positive — a claim that the `go.yml` `push` trigger should target `main`, when the repository default branch and this PR's base are both `master`. The re-review came back with no further comments.

**Known deferrals** (not addressed in this PR):

- No test that `go vet -copylocks` actually catches a copy of `fifomu.Mutex` (would require a `testscript`-style exec, low criticality).
- No property-based/fuzz test beyond the chaos test.
- `synctest` not adopted for deterministic timing tests — the package-global `waiterPool` caches channels across tests, and `synctest` panics on cross-bubble channel ops; working around it wasn't worth the contortion.

---

## Commits

```
fb7e9a0 test+docs: address Copilot review — deterministic enqueue waits, dep label
4389b05 ci: migrate golangci-lint to v2, bump actions, dedupe PR triggers
7fc96a1 test: relax flaky cancel/acquire outcome assertion; add deadlock guard
40e67e1 Pre-review polish: gitignore .claude, fix stale benchmark name
7a2b800 Polish: rename _Uncontended benchmark, tighten chaos assertions
a085586 docs: README corrections from review round 2
6608aad test: recover-safety, concurrent double-unlock, remove field-clearing
1bbed0e Code review round 2: core fixes
8fbf040 docs: clarity pass — add When to use, How it works, Testing, Requirements
12df6f8 bench: add LockContext benchmarks; refresh all numbers for Go 1.26
94fe197 test: add chaos/mutual-exclusion stress test
a8d1905 test: add goleak to catch stray goroutines across all tests
6cbc140 waiter: add signal/wait/tryDrain methods (tryDrain later renamed to tryReceive)
5449348 list: drop generics; collapse to concrete waiter types
80d7c08 list: rename Value to value, len uint to int
286089c Add explicit noCopy field to Mutex
a523051 refactor: collapse cur int64 to locked bool
a15daff Add correctness tests: mixed queue, middle-of-queue cancel, invariants
a827434 Address PR review feedback: self-enforcing invariant, deterministic FIFO tests, comment polish
8e4f1a8 Fix LockContext cancellation deadlock; harden queue; upgrade to Go 1.26
```

## Test plan

- [x] `go vet ./...` — clean
- [x] `go test -race -count=3 ./...` — all 27 tests pass
- [x] Regression test verified to deadlock within seconds when the buffer change is reverted
- [x] Regression test verified to panic cleanly (not deadlock) when the send-with-default guard is violated
- [x] Chaos test: 137k+ Lock ops in 2s, zero mutual-exclusion violations
- [x] `goleak` guarding every test; no stranded goroutines
- [x] Benchmarks re-run on Go 1.26 / M1 Max, results match the README
- [x] Downstream `go vet -copylocks` verified manually to flag `fifomu.Mutex` copies via the explicit `noCopy`

## Scope notes

- Binary ABI unchanged; `fifomu.Mutex` struct layout is the same size class.
- `Lock` / `Unlock` / `TryLock` callers see no behavior change under correct use.
- `LockContext` callers see strictly more documented cancel semantics (no deadlock). The cancel-after-acquire → nil behavior is unchanged.
- `Unlock` post-panic semantics are **improved**: after a recovered double-unlock panic, the mutex remains usable (vs. `sync.Mutex` which leaves `state < 0`).
- `go.mod` now requires Go 1.26. Only the test binary uses post-1.20 features.

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -5,22 +5,34 @@ name: Go
 
 on:
   push:
+    branches: [master]
```
[the remark this answers is not in the document]
Comment (neilotoole) [reply .github/workflows/go.yml:8]: This is a false positive: the repo's default branch and this PR's base are both `master`, not `main` (confirmed via `gh repo view --json defaultBranchRef` and the PR's `baseRefName`). So `push.branches: [master]` does fire on the default branch. Feature-branch pushes are covered by the `pull_request` event; restricting `push` to `master` is precisely what removes the duplicate (push + pull_request) run sets this PR was getting. Leaving as-is.

File: fifomu_context_test.go [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- mu.LockContext(ctx)
+	}()
+
+	// Let the goroutine queue itself.
+	time.Sleep(20 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (neilotoole) [reply fifomu_context_test.go]: Fixed in fb7e9a0 — replaced the fixed `time.Sleep` with the deterministic `waitForWaiters(t, &mu, 1)` gate.

File: fifomu_context_test.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+	defer cancel(nil)
+
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- mu.LockContext(ctx)
+	}()
+
+	time.Sleep(20 * time.Millisecond)
+
+	cancel(want)
```
[the remark this answers is not in the document]
Comment (neilotoole) [reply fifomu_context_test.go:96]: Fixed in fb7e9a0 — same change here; now gates on `waitForWaiters(t, &mu, 1)` instead of a fixed sleep.

File: README.md [hunk 4]
```diff
@@ -82,46 +189,93 @@ The benchmarks compare three mutex implementations:
...
+BenchmarkMutexSpin/fifomu-10                  493010       2404     ns/op    0 B/op   0 allocs/op
+BenchmarkMutexSpin/semaphoreMu-10             478358       2511     ns/op  175 B/op   2 allocs/op
 ```
 
+## Requirements
+
+- **Go 1.26 or later.** The package is generic-free, but the tests and
+  benchmarks use `for b.Loop()` (Go 1.24+) and `sync.WaitGroup.Go`
+  (Go 1.25+).
+- **Runtime dependencies:** [`golang.org/x/sync`](https://pkg.go.dev/golang.org/x/sync),
```
[the remark this answers is not in the document]
Comment (neilotoole) [reply README.md]: Fixed in fb7e9a0 — relabeled to "Non-stdlib dependency (tests only)" so it no longer implies `x/sync` is a runtime dependency.
